File: triskell_command/views/funnel.py
"""Vue Funnel — visualise la conversion Prospects → Envoyés → Réponses →
Intéressés → Gagnés, par segment et par période.

Lit funnel_metrics qui agrège prospects + email_history. Pas de tracking
externe : c'est une vue purement basée sur les données déjà en Supabase.
"""

from __future__ import annotations

import logging

import customtkinter as ctk

from .. import theme as T
from ..integrations import funnel_metrics
from ..widgets.components import (
    Card,
    Chip,
    EmptyState,
    SecondaryButton,
    StatCard,
    ViewHeader,
)
from .base import BaseView

logger = logging.getLogger(__name__)


PERIOD_LABELS = {
    "7d":  "7 jours",
    "30d": "30 jours",
    "90d": "90 jours",
    "all": "Tout",
}
SEGMENT_LABELS = {
    "all":       "Tous",
    "creators":  "Créateurs",
    "b2b_local": "B2B local",
}
STAGE_LABELS = [
    ("prospects",  "Prospects"),
    ("sent",       "Envoyés"),
    ("replies",    "Réponses"),
    ("interested", "Intéressés"),
    ("won",        "Gagnés"),
]


class FunnelView(BaseView):
    title = "Conversions"
    subtitle = (
        "Tes taux de transformation par type de prospect et par période. "
        "Tu vois en un coup d'œil ce qui marche et ce qui coince."
    )

    def build(self) -> None:
        c = self.colors
        header = ViewHeader(self, title=self.title, subtitle=self.subtitle, colors=c)
        header.pack(fill="x", padx=T.SPACE_2XL, pady=(T.SPACE_LG, T.SPACE_MD))

        SecondaryButton(header.actions, colors=c, icon="refresh",
                         text="Rafraîchir",
                         command=self._refresh).pack(side="left")

        # Filtres en deux lignes pour respirer
        filters_wrap = ctk.CTkFrame(self, fg_color="transparent")
        filters_wrap.pack(fill="x", padx=T.SPACE_2XL, pady=(T.SPACE_SM, T.SPACE_LG))

        # Période
        period_block = ctk.CTkFrame(filters_wrap, fg_color="transparent")
        period_block.pack(fill="x", anchor="w", pady=(0, T.SPACE_SM))
        ctk.CTkLabel(
            period_block, text="PÉRIODE",
            font=(T.FONT_FAMILY_FALLBACK, T.FONT_SIZE_TINY, "bold"),
            text_color=c.text_muted,
        ).pack(side="left", padx=(0, T.SPACE_MD))
        self._period = "30d"
        self._period_chips: dict[str, Chip] = {}
        for k, label in PERIOD_LABELS.items():
            chip = Chip(period_block, text=label, colors=c,
                         is_active=(k == self._period),
                         on_toggle=lambda _v, kk=k: self._set_period(kk))
            chip.pack(side="left", padx=(0, T.SPACE_SM))
            self._period_chips[k] = chip

        # Segment (sur sa propre ligne)
        seg_block = ctk.CTkFrame(filters_wrap, fg_color="transparent")
        seg_block.pack(fill="x", anchor="w")
        ctk.CTkLabel(
            seg_block, text="TYPE DE PROSPECT",
            font=(T.FONT_FAMILY_FALLBACK, T.FONT_SIZE_TINY, "bold"),
            text_color=c.text_muted,
        ).pack(side="left", padx=(0, T.SPACE_MD))
        self._segment = "all"
        self._seg_chips: dict[str, Chip] = {}
        for k, label in SEGMENT_LABELS.items():
            chip = Chip(seg_block, text=label, colors=c,
                         is_active=(k == self._segment),
                         on_toggle=lambda _v, kk=k: self._set_segment(kk))
            chip.pack(side="left", padx=(0, T.SPACE_SM))
            self._seg_chips[k] = chip

        # Container scroll
        self._scroll = ctk.CTkScrollableFrame(
            self, fg_color="transparent",
            scrollbar_button_color=c.border_strong,
        )
        self._scroll.pack(fill="both", expand=True,
                          padx=T.SPACE_2XL, pady=(0, T.SPACE_LG))

        self._status_var = ctk.StringVar(value="")
        ctk.CTkLabel(
            self, textvariable=self._status_var,
            font=(T.FONT_FAMILY_FALLBACK, T.FONT_SIZE_SMALL),
            text_color=c.text_muted,
        ).pack(fill="x", padx=T.SPACE_2XL, pady=(0, T.SPACE_SM))

    def on_show(self) -> None:
        self._refresh()

    def _set_period(self, period: str) -> None:
        self._period = period
        for k, chip in self._period_chips.items():
            chip.set_active(k == period)
        self._refresh()

    def _set_segment(self, segment: str) -> None:
        self._segment = segment
        for k, chip in self._seg_chips.items():
            chip.set_active(k == segment)
        self._refresh()

    def _refresh(self) -> None:
        for w in self._scroll.winfo_children():
            w.destroy()
        c = self.colors

        data = funnel_metrics.compute_funnel(period=self._period,
                                              segment=self._segment)
        if not data.get("ok"):
            EmptyState(
                self._scroll, colors=c, icon="settings",
                title="Connexion à la base partagée requise",
                message="Les chiffres de conversion viennent de la base "
                        "partagée Triskell. Connecte-toi depuis les Réglages.",
                cta_text="Aller dans Réglages",
                cta_command=lambda: self._navigate_to("config"),
            ).pack(fill="both", expand=True)
            self._status_var.set("Connexion requise.")
            return

        stages = data["stages"]

        # ---- Funnel principal (5 cartes) ----
        grid = ctk.CTkFrame(self._scroll, fg_color="transparent")
        grid.pack(fill="x", pady=(0, T.SPACE_LG))
        for col in range(len(STAGE_LABELS)):
            grid.grid_columnconfigure(col, weight=1, uniform="funnel")

        prev = None
        for i, (key, label) in enumerate(STAGE_LABELS):
            value = stages.get(key, 0)
            delta = ""
            accent = ""
            if prev is not None and prev > 0:
                pct = round(100 * value / prev, 1)
                delta = f"{pct}% du précédent"
                if key == "won" and pct >= 5:
                    accent = c.success
                elif key == "interested" and pct >= 5:
                    accent = c.success
            elif key == "prospects":
                delta = f"sur {self._period}"
            StatCard(
                grid, label=label, value=str(value),
                delta=delta, accent=accent, colors=c,
            ).grid(row=0, column=i,
                    padx=(0 if i == 0 else T.SPACE_SM,
                            0 if i == len(STAGE_LABELS) - 1 else T.SPACE_SM),
                    sticky="ew")
            prev = value

        # ---- Répartition par catégorie de réponse ----
        by_cat = data.get("by_category") or {}
        if by_cat:
            self._section_label("Types de réponses reçues")
            self._horizontal_bar(by_cat, color=c.accent_secondary)

        # ---- Répartition par statut prospect ----
        by_status = data.get("by_status") or {}
        if by_status:
            self._section_label("Où en sont tes prospects")
            self._horizontal_bar(by_status, color=c.accent)

        # ---- Top produits / templates ----
        by_prod = data.get("by_product") or {}
        if by_prod:
            self._section_label("Produits les plus mis en avant")
            self._horizontal_bar(by_prod, color=c.text_secondary)

        # Status bar
        sent = stages["sent"]
        won = stages["won"]
        interested = stages["interested"]
        denom = max(sent, 1)
        self._status_var.set(
            f"Période : {PERIOD_LABELS.get(self._period)} · "
            f"Type : {SEGMENT_LABELS.get(self._segment)} · "
            f"Taux de réponse {round(100*stages['replies']/denom, 1)}% · "
            f"Taux d'intérêt {round(100*interested/denom, 1)}% · "
            f"Taux de gain {round(100*won/max(stages['prospects'],1), 1)}%"
        )

    def _section_label(self, text: str) -> None:
        c = self.colors
        wrap = ctk.CTkFrame(self._scroll, fg_color="transparent")
        wrap.pack(fill="x", pady=(T.SPACE_2XL, T.SPACE_MD))
        bar = ctk.CTkFrame(wrap, fg_color=c.accent, width=24, height=2,
                            corner_radius=1)
        bar.pack(side="left", padx=(0, T.SPACE_SM))
        bar.pack_propagate(False)
        ctk.CTkLabel(
            wrap, text=text.upper(),
            font=(T.FONT_FAMILY_FALLBACK, T.FONT_SIZE_TINY, "bold"),
            text_color=c.text_muted, anchor="w",
        ).pack(side="left", fill="x")

    def _horizontal_bar(self, data: dict, *, color: str) -> None:
        """Petit graphe : labels + barres proportionnelles."""
        c = self.colors
        if not data:
            return
        max_val = max(data.values()) or 1
        card = Card(self._scroll, colors=c)
        card.pack(fill="x", pady=(0, T.SPACE_MD))
        inner = ctk.CTkFrame(card, fg_color="transparent")
        inner.pack(fill="x", padx=T.SPACE_LG, pady=T.SPACE_LG)
        for k, v in sorted(data.items(), key=lambda kv: -kv[1]):
            row = ctk.CTkFrame(inner, fg_color="transparent")
            row.pack(fill="x", pady=(0, 2))
            ctk.CTkLabel(
                row, text=str(k),
                font=(T.FONT_FAMILY_FALLBACK, T.FONT_SIZE_SMALL),
                text_color=c.text_secondary, anchor="w", width=180,
            ).pack(side="left")
            # Barre
            bar_wrap = ctk.CTkFrame(row, fg_color=c.panel,
                                     corner_radius=T.RADIUS_SM, height=14)
            bar_wrap.pack(side="left", fill="x", expand=True,
                           padx=(T.SPACE_SM, T.SPACE_SM))
            bar_wrap.pack_propagate(False)
            try:
                pct = max(2, int(100 * v / max_val))
            except Exception:
                pct = 2
            inner_bar = ctk.CTkFrame(
                bar_wrap, fg_color=color,
                width=int(pct * 4), height=14, corner_radius=T.RADIUS_SM,
            )
            inner_bar.pack(side="left")
            inner_bar.pack_propagate(False)
            ctk.CTkLabel(
                row, text=str(v),
                font=(T.FONT_FAMILY_FALLBACK, T.FONT_SIZE_SMALL, "bold"),
                text_color=c.text_primary, width=40,
            ).pack(side="left")

File: triskell_command/web/ui/scripts/funnel.js
/* Vue Conversions — funnel par segment / période */

const Funnel = {
  period: '30d',
  segment: 'all',
  PERIODS: { '7d': '7 jours', '30d': '30 jours', '90d': '90 jours', 'all': 'Tout' },
  // Libellés longs pour les titres de sections (« 7 derniers jours »…)
  PERIOD_TITLES: { '7d': '7 derniers jours', '30d': '30 derniers jours', '90d': '90 derniers jours', 'all': 'depuis le début' },
  SEGMENTS: { 'all': 'Tous', 'creators': 'Créateurs', 'b2b_local': 'Commerces locaux' },
  STAGES: [
    { key: 'prospects',  label: 'Prospects',  sub: 'Contacts entrés dans ta base sur la période' },
    { key: 'sent',       label: 'Envoyés',    sub: 'Prospects qui ont reçu au moins un mail' },
    { key: 'replies',    label: 'Réponses',   sub: 'Prospects qui ont répondu (peu importe quoi)' },
    { key: 'interested', label: 'Intéressés', sub: 'Ont répondu « ça m’intéresse » — à toi de jouer' },
    { key: 'won',        label: 'Gagnés',     sub: 'Devenus clients : projet créé ou vente encaissée' },
  ],

  async render(container) {
    container.innerHTML = `
      <section class="animate-slide-up">
        <div class="mb-6 sm:mb-8">
          <div class="flex items-start justify-between gap-3">
            <div class="min-w-0 flex-1">
              <div class="hero-kicker mb-2">CONVERSIONS</div>
              <h1 class="hero-title hero-title--md mb-2 sm:mb-3">Tes taux de transformation.</h1>
              <p class="hero-subtitle">Tu vois en un coup d'œil ce qui marche et ce qui coince.</p>
            </div>
            ${Help.button('funnel')}
          </div>
        </div>

        <div id="f-filters" class="space-y-3 sm:space-y-4 mb-6 sm:mb-10"></div>
        <div id="f-content"></div>
      </section>
    `;
    this._renderFilters();
    await this.refresh();
  },

  _renderFilters() {
    const wrap = document.getElementById('f-filters');
    if (!wrap) return;
    wrap.innerHTML = `
      <div>
        <div class="text-[11px] font-bold tracking-widest text-text-muted mb-2">PÉRIODE</div>
        <div class="flex gap-2 flex-wrap" id="f-periods"></div>
      </div>
      <div>
        <div class="text-[11px] font-bold tracking-widest text-text-muted mb-2">TYPE DE PROSPECT</div>
        <div class="flex gap-2 flex-wrap" id="f-segments"></div>
      </div>
    `;
    const periods = document.getElementById('f-periods');
    Object.entries(this.PERIODS).forEach(([key, label]) => {
      const active = key === this.period;
      const btn = document.createElement('button');
      btn.textContent = label;
      btn.style.cssText = this._chipStyle(active);
      btn.onclick = () => { this.period = key; this._renderFilters(); this.refresh(); };
      periods.appendChild(btn);
    });
    const segs = document.getElementById('f-segments');
    Object.entries(this.SEGMENTS).forEach(([key, label]) => {
      const active = key === this.segment;
      const btn = document.createElement('button');
      btn.textContent = label;
      btn.style.cssText = this._chipStyle(active);
      btn.onclick = () => { this.segment = key; this._renderFilters(); this.refresh(); };
      segs.appendChild(btn);
    });
  },

  _chipStyle(active) {
    if (active) {
      return `padding: 7px 14px; border-radius: 999px; font-size: 12.5px;
              font-weight: 600; background: hsl(var(--accent-strong));
              color: hsl(var(--on-accent)); border: 1px solid hsl(var(--accent-strong));
              cursor: pointer; transition: all 160ms;`;
    }
    return `padding: 7px 14px; border-radius: 999px; font-size: 12.5px;
            font-weight: 500; background: transparent;
            color: hsl(var(--text-secondary));
            border: 1px solid hsl(var(--border-strong)); cursor: pointer;
            transition: all 160ms;`;
  },

  async refresh() {
    const slot = document.getElementById('f-content');
    if (!App.api) {
      slot.innerHTML = this._preview();
      return;
    }
    slot.innerHTML = `<div class="text-center py-12 text-text-muted">Chargement…</div>`;
    let data;
    try {
      data = await App.api.get_funnel({ period: this.period, segment: this.segment });
    } catch (e) {
      console.error('[Funnel] chargement', e);
      slot.innerHTML = `
        <div class="card p-10 text-center">
          <div class="text-3xl mb-3">⚠️</div>
          <h2 class="text-xl font-semibold mb-2">Impossible de charger tes conversions</h2>
          <p class="text-text-secondary mb-6">Le serveur n'a pas répondu. Vérifie ta connexion, puis réessaie.</p>
          <button class="btn btn-primary" onclick="Funnel.refresh()">Réessayer</button>
        </div>
      `;
      return;
    }
    if (!data || !data.ok) {
      if (data && data.error) console.warn('[Funnel] connexion', data.error);
      slot.innerHTML = `
        <div class="card p-10 text-center">
          <div class="text-3xl mb-3">🔌</div>
          <h2 class="text-xl font-semibold mb-2">Connexion requise</h2>
          <p class="text-text-secondary mb-6">Connecte-toi à la base partagée Triskell.</p>
          <button class="btn btn-primary" onclick="App.show('config', {tab:'account'})">Aller dans Réglages</button>
        </div>
      `;
      return;
    }
    // Tout à zéro sur la période → on ORIENTE au lieu d'aligner des 0 muets
    const stagesTotal = Object.values(data.stages || {}).reduce((a, b) => a + (b || 0), 0);
    const emptyNote = stagesTotal === 0 ? `
      <div class="card p-8 text-center text-text-muted mb-10 -mt-6">
        <div class="text-3xl mb-3">🌱</div>
        <p class="text-sm max-w-xl mx-auto">
          Rien à mesurer sur cette période pour l'instant. Les chiffres apparaissent dès que
          des prospects entrent dans la base et que des mails partent —
          lance une prospection pour démarrer, ou élargis la période ci-dessus.
        </p>
      </div>
    ` : '';
    slot.innerHTML = this._renderStages(data) +
                     emptyNote +
                     `<div id="f-templates"></div>` +
                     this._renderBars('Types de réponses reçues', data.by_category, 'accent') +
                     this._renderBars('Où en sont tes prospects', data.by_status, 'accent-glow') +
                     this._renderBars('Produits les plus mis en avant', data.by_product, 'text-secondary');
    this._loadTemplatePerf();
  },

  // ---- Performance par modèle de mail (chargée à part : requête lourde) ----
  async _loadTemplatePerf() {
    // (Pas de garde typeof sur App.api.funnel_by_template : App.api est un
    // Proxy, typeof renvoie toujours "function" — la garde ne testait rien.)
    if (!App.api || !document.getElementById('f-templates')) return;
    // Le tableau suit le MÊME filtre de période que le reste de l'écran
    const wantedPeriod = this.period;
    let r = null;
    try {
      r = await App.api.funnel_by_template({ period: wantedPeriod });
    } catch (e) {
      console.warn('[Funnel] perf par modèle', e);
      r = null;
    }
    // L'utilisateur a changé de période pendant le chargement → résultat périmé
    if (wantedPeriod !== this.period) return;
    // On relit le créneau APRÈS l'attente : l'écran a pu être re-rendu entre-temps
    const slot = document.getElementById('f-templates');
    if (!slot) return;
    if (!r || !r.ok || !(r.rows || []).length) {
      slot.innerHTML = '';
      return;
    }
    const rows = r.rows.slice(0, 12);
    slot.innerHTML = `
      <div class="mb-10">
        <div class="section-label">Quel modèle de mail fait répondre ? (${this.PERIOD_TITLES[wantedPeriod] || wantedPeriod})</div>
        <div class="card p-0 overflow-hidden">
          <table class="w-full text-sm">
            <thead class="bg-bg">
              <tr class="text-left text-[11px] uppercase tracking-wide text-text-muted">
                <th class="px-4 py-2.5">Modèle</th>
                <th class="px-4 py-2.5 text-right">Envoyés</th>
                <th class="px-4 py-2.5 text-right">Réponses</th>
                <th class="px-4 py-2.5 text-right">Intéressés</th>
                <th class="px-4 py-2.5 text-right">Taux de réponse</th>
              </tr>
            </thead>
            <tbody>
              ${rows.map(t => `
                <tr class="border-t border-border">
                  <td class="px-4 py-2.5 font-medium truncate max-w-[280px]" title="${this._esc(t.template)}">${this._esc(t.template)}</td>
                  <td class="px-4 py-2.5 text-right tabular-nums">${t.sent}</td>
                  <td class="px-4 py-2.5 text-right tabular-nums">${t.replies}</td>
                  <td class="px-4 py-2.5 text-right tabular-nums ${t.interested > 0 ? 'text-success-text font-semibold' : ''}">${t.interested}</td>
                  <td class="px-4 py-2.5 text-right tabular-nums font-bold ${t.reply_rate >= 5 ? 'text-success-text' : ''}">${t.reply_rate}%</td>
                </tr>`).join('')}
            </tbody>
          </table>
          <div class="px-4 py-2.5 text-[11px] text-text-muted border-t border-border">
            Une réponse est créditée au dernier modèle envoyé à ce prospect.
            Moins de 20 envois par modèle = chiffres à prendre avec des pincettes.
          </div>
        </div>
      </div>`;
  },

  _renderStages(data) {
    const stages = data.stages || {};
    let prev = null;
    return `
      <div class="grid grid-cols-2 sm:grid-cols-3 lg:grid-cols-5 gap-4 mb-4">
        ${this.STAGES.map((s, i) => {
          const value = stages[s.key] || 0;
          let delta = '';
          if (prev !== null && prev > 0) {
            const pct = Math.round(100 * value / prev * 10) / 10;
            delta = `${pct}% du précédent`;
          } else if (i === 0) {
            delta = `sur ${this.PERIODS[this.period]}`;
          }
          const accent = (s.key === 'won' && prev > 0 && (value/prev) >= 0.05) ? 'success' :
                         (s.key === 'interested' && prev > 0 && (value/prev) >= 0.05) ? 'success' : '';
          prev = value;
          return `
            <div class="stat-card ${accent ? 'accent-' + accent : ''}" title="${this._esc(s.sub || '')}">
              <div class="label">${s.label}</div>
              <div class="value">${value}</div>
              ${delta ? `<div class="delta">${delta}</div>` : ''}
            </div>
          `;
        }).join('')}
      </div>
      <div class="text-xs text-text-muted mb-12" style="text-wrap: pretty">
        <b>Prospects</b> = entrés dans ta base · <b>Envoyés</b> = ont reçu un mail ·
        <b>Réponses</b> = ont répondu · <b>Intéressés</b> = ont dit « ça m’intéresse » ·
        <b>Gagnés</b> = devenus clients.
        Repère : <b>2 à 5 réponses pour 100 mails</b> envoyés, c'est la norme en
        prospection à froid.
      </div>
    `;
  },

  _renderBars(title, data, color) {
    if (!data || Object.keys(data).length === 0) return '';
    const max = Math.max(...Object.values(data));
    const sorted = Object.entries(data).sort((a, b) => b[1] - a[1]);
    return `
      <div class="mb-10">
        <div class="section-label">${this._esc(title)}</div>
        <div class="card p-6 space-y-3">
          ${sorted.map(([k, v]) => {
            const pct = Math.max(2, Math.round(100 * v / max));
            return `
              <div class="flex items-center gap-4">
                <div class="w-28 sm:w-44 text-sm text-text-secondary truncate" title="${this._esc(k)}">${this._esc(k)}</div>
                <div class="flex-1 h-2 rounded-full" style="background: hsl(var(--border));">
                  <div class="h-full rounded-full" style="width: ${pct}%; background: hsl(var(--${color === 'text-secondary' ? 'text-secondary' : color}));"></div>
                </div>
                <div class="w-12 text-right text-sm font-bold">${v}</div>
              </div>
            `;
          }).join('')}
        </div>
      </div>
    `;
  },

  _esc(s) {
    return String(s || '').replace(/[&<>"']/g, c => ({
      '&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'
    }[c]));
  },

  _preview() {
    const fake = {
      ok: true,
      stages: { prospects: 1240, sent: 312, replies: 28, interested: 8, won: 3 },
      by_category: { interested: 8, not_now: 6, no: 9, unsubscribe: 3, unknown: 2 },
      by_status: { new: 870, qualified: 180, contacted: 124, replied: 38, won: 3 },
      by_product: { 'obelisk': 84, 'pack-elec': 62, 'eliks': 41, 'studio-pdf': 24 },
    };
    const banner = `
      <div class="card p-3 mb-6 text-center text-[12px] text-text-muted">
        Données d'exemple — connecte l'app au serveur pour voir tes vrais chiffres.
      </div>
    `;
    return banner +
           this._renderStages(fake) +
           this._renderBars('Types de réponses reçues', fake.by_category, 'accent') +
           this._renderBars('Où en sont tes prospects', fake.by_status, 'accent-glow') +
           this._renderBars('Produits les plus mis en avant', fake.by_product, 'text-secondary');
  },
};
